### PR TITLE
[Sprint 22/23 / PD-405] [Feature] Add implementation GRPC

### PR DESCRIPTION
**Pull request by mbsaloka**
## Jira Link: 
https://ichiro-its.atlassian.net/browse/PD-405?atlOrigin=eyJpIjoiZjcyMmNiZGY0YjQ5NDliYzkxYjhkZjBiODFmODk2OWUiLCJwIjoiaiJ9

## Description

Added grpc to implement get, set and save functions. The gRPC server runs on port :5858 and consists of 3 RPCs, GetColorSetting, SetColorSetting, and SaveColorSetting. It has been tested using Postman and it works.

Example Message:
> GetColorSetting
```
{}
```

> SetColorSetting
```
{
    "max_hue": 123,
    "max_saturation": 123,
    "max_value": 123,
    "min_hue": 123,
    "min_saturation": 123,
    "min_value": 123,
    "name": "line"
}
```

> SaveColorSetting
```
{
    "json_color": "{\"ball\":{\"max_hsv\":[179,100,100],\"min_hsv\":[151,40,10]},\"field\":{\"max_hsv\":[116,100,100],\"min_hsv\":[84,10,0]},\"line\":{\"max_hsv\":[260,60,100],\"min_hsv\":[180,0,55]}}"
}
```

## Type of Change

- [ ] Bugfix
- [ ] Enhancement
- [x] New feature
- [ ] Breaking change (fix or feature that would cause the existing functionality to not work as expected)

## How Has This Been Tested?

- [ ] New unit tests added.
- [x] Manual tested.

## Checklist:

- [x] Using Branch Name Convention
    - `feature/JIRA-ID-SHORT-DESCRIPTION` if has a JIRA ticket
    - `enhancement/SHORT-DESCRIPTION` if has/has no JIRA ticket and contain enhancement
    - `hotfix/SHORT-DESCRIPTION` if the change doesn't need to be tested (urgent)
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made the documentation for the corresponding changes.

### CODE REVIEW

**Review by FaaizHaikal — CHANGES_REQUESTED** [review]
remove ```.vscode``` dir

File: include/ninshiki_cpp/config/grpc/config.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+  static inline std::unique_ptr<grpc::ServerCompletionQueue> cq_;
+  static inline std::unique_ptr<grpc::Server> server_;
+  std::thread thread_;
+  ninshiki_interfaces::proto::Config::AsyncService service_;
+};
+
+}  // namespace ninshiki_cpp
+
+#endif  // NINSHIKI_CPP__CONFIG__GRPC__CONFIG_HPP_
```
Comment (FaaizHaikal) [inline include/ninshiki_cpp/config/grpc/config.hpp:72]: add ```EOF``` line. do this for all files

File: src/ninshiki_cpp/detector/color_detector.cpp [hunk 2]
```diff
@@ -132,6 +132,30 @@ bool ColorDetector::sync_configuration()
...
+    if (item.name == color.name) {
+      item.min_hue = color.min_hue;
+      item.max_hue = color.max_hue;
+      item.min_saturation = color.min_saturation;
+      item.max_saturation = color.max_saturation;
+      item.min_value = color.min_value;
+      item.max_value = color.max_value;
+
+      std::cout << "[DEBUG] Color setting has been applied!" << std::endl;
+      std::cout << "name: " << item.name << std::endl;
```
Comment (FaaizHaikal) [inline src/ninshiki_cpp/detector/color_detector.cpp:147]: remove these debugging lines

**Review by FaaizHaikal — CHANGES_REQUESTED** [review]

**Review by threeal — CHANGES_REQUESTED** [review]
I suggest creating a dedicated gRPC utility package first before merging this change, as I see there are many duplications of `CallData`, `CallDataBase`, etc., across different projects that are going to implement gRPC (see https://github.com/ichiro-its/shisen_cpp/pull/25 and https://github.com/ichiro-its/aruku/pull/20).

File: data/color_classifier.json [hunk 3]
```diff
@@ -0,0 +1,38 @@
+{
```
Comment (borednuna) [inline data/color_classifier.json:1]: We can remove this as we have moved the configurations to config repo

File: include/ninshiki_cpp/utils/circles.hpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2021 ICHIRO ITS
```
Comment (borednuna) [inline include/ninshiki_cpp/utils/circles.hpp:1]: Isn't this done in branch refactor ninshiki ? Maybe remove all changes that are done in refactor ninshiki branch before merging this branch later

File: include/ninshiki_cpp/config/grpc/call_data_get_color_setting.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+{
+public:
+  CallDataGetColorSetting(
+    ninshiki_interfaces::proto::Config::AsyncService * service, grpc::ServerCompletionQueue * cq,
+    const std::string & path);
+
+protected:
+  void AddNextToCompletionQueue() override;
+  void WaitForRequest();
+  void HandleRequest();
```
Comment (threeal) [inline include/ninshiki_cpp/config/grpc/call_data_get_color_setting.hpp:39]: These lines should be marked as override.
```suggestion
  void WaitForRequest() override;
  void HandleRequest() override;
```

File: include/ninshiki_cpp/config/grpc/call_data_save_color_setting.hpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+{
+public:
+  CallDataSaveColorSetting(
+    ninshiki_interfaces::proto::Config::AsyncService * service, grpc::ServerCompletionQueue * cq,
+    const std::string & path);
+
+protected:
+  void AddNextToCompletionQueue() override;
+  void WaitForRequest();
+  void HandleRequest();
```
Comment (threeal) [inline include/ninshiki_cpp/config/grpc/call_data_save_color_setting.hpp:40]: Ditto.

File: include/ninshiki_cpp/config/grpc/call_data_set_color_setting.hpp [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+public:
+  CallDataSetColorSetting(
+    ninshiki_interfaces::proto::Config::AsyncService * service, grpc::ServerCompletionQueue * cq,
+    const std::string & path, std::shared_ptr<ninshiki_cpp::detector::ColorDetector> color_detection);
+    using ColorDetector = ninshiki_cpp::detector::ColorDetector;
+
+protected:
+  void AddNextToCompletionQueue() override;
+  void WaitForRequest();
+  void HandleRequest();
```
Comment (threeal) [inline include/ninshiki_cpp/config/grpc/call_data_set_color_setting.hpp:43]: Ditto.

File: include/ninshiki_cpp/config/grpc/config.hpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+#include <rclcpp/rclcpp.hpp>
+
+#include <chrono>
+#include <fstream>
+#include <future>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <thread>
```
Comment (threeal) [inline include/ninshiki_cpp/config/grpc/config.hpp:45]: Some of these headers may be unnecessary, don't put all of them in the header file.

File: src/ninshiki_cpp/config/grpc/call_data_base.cpp [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "ninshiki_cpp/config/grpc/call_data_base.hpp"
+
+namespace ninshiki_cpp
+{
+CallDataBase::CallDataBase() {}
```
Comment (threeal) [inline src/ninshiki_cpp/config/grpc/call_data_base.cpp:25]: Just remove the constructor if it does not do anything.

File: src/ninshiki_cpp/config/grpc/call_data_set_color_setting.cpp [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+    utils::Color color(
+      name,
+      min_hue,
+      max_hue,
+      min_saturation,
+      max_saturation,
+      min_value,
+      max_value
+    );
```
Comment (threeal) [inline src/ninshiki_cpp/config/grpc/call_data_set_color_setting.cpp:67]: Simpler
```suggestion
    utils::Color color(
      request_.name(),
      request_.min_hue(),
      request_.max_hue(),
      request_.min_saturation(),
      request_.max_saturation(),
      request_.min_value(),
      request_.max_value()
    );
```

**Review by threeal — CHANGES_REQUESTED** [review]

File: src/ninshiki_cpp/config/grpc/call_data_base.cpp (hunk 9 above)
Comment (mbsaloka) [reply src/ninshiki_cpp/config/grpc/call_data_base.cpp:25]: after I deleted the constructor or deleted the `call_data_base.cpp` file, when building the package it produced an error
```
/usr/bin/ld: libninshiki_cpp.so: undefined reference to `ninshiki_cpp::CallDataBase::CallDataBase()'
```

Comment (threeal) [reply src/ninshiki_cpp/config/grpc/call_data_base.cpp:25]: You also need to remove the constructor from the class declaration in the `.hpp` file.

**Review by FaaizHaikal — APPROVED** [review]

**Comment by FaaizHaikal** [comment]
> I suggest creating a dedicated gRPC utility package first before merging this change, as I see there are many duplications of `CallData`, `CallDataBase`, etc., across different projects that are going to implement gRPC (see [ichiro-its/shisen_cpp#25](https://github.com/ichiro-its/shisen_cpp/pull/25) and [ichiro-its/aruku#20](https://github.com/ichiro-its/aruku/pull/20)).

I think there are going to be a lot of changes while working on this. It might be worked on after this year's competition mas, which is around next month.

**Pull request closed without merging**